src/qpx_mmc.cpp

```diff
@@ -0,0 +1,4244 @@
+
+  drive->cmd[6] = (len >> 16) & 0xFF;
+  drive->cmd[7] = (len >> 8) & 0xFF;
+  drive->cmd[8] = (len)&0xFF;
+  drive->cmd[9] = 0;
+
+  if ((drive->err = drive->cmd.transport(READ, drive->rd_buf, len))) {
+    if (!drive->silent)
+      sperror("READ_BUFFER", drive->err);
+    return (drive->err);
+  }
+
+  return 0;
+}
+
+int read_echo_buffer_size(drive_info *drive) {
+  drive->cmd[0] = SPC_READ_BUFFER;
+  //	drive->cmd[1] = 0x00;
+  drive->cmd[1] = 0x0B; // echo buffer descriptor
+  drive->cmd[8] = 4;
+  drive->cmd[9] = 0;
+
+  if ((drive->err = drive->cmd.transport(READ, drive->rd_buf, 4))) {
+    if (!drive->silent)
+      sperror("READ_ECHO_BUFFER_SIZE", drive->err);
+    return (0);
+  }
+
+  return ntoh16(drive->rd_buf + 2) & 0x1FFF;
+}
+
+int test_dma_speed(drive_info *drive, long msecs) {
+  long st, et;
+  long speed = 0;
+
+  int cnt = 0;
+  drive->silent--;
+
+  uint32_t buf_size = read_echo_buffer_size(drive);
+  std::cout << "Echo buffer size: " << buf_size << "\n";
+
+  buf_size = 4096;
+
+  if (!buf_size) {
+    drive->silent++;
+    return 0;
+  }
+
+  if (!drive->silent)
+    printf("** Testing DMA speed...\n");
+  st = getmsecs();
+  for (et = getmsecs(); (et - st) < (msecs); et = getmsecs()) {
+    if (write_buffer(drive, 0x0A, 0, 0, buf_size)) {
+      if (!drive->silent)
+        printf("WRITE BUFFER error! DMA speed test aborted!\n");
+      drive->silent++;
+      return 1;
+    }
+    if (read_buffer(drive, 0x0A, 0, 0, buf_size)) {
+      if (!drive->silent)
+        printf("READ BUFFER error! DMA speed test aborted!\n");
+      drive->silent++;
+      return 1;
+    }
+    cnt++;
+  }
+  speed = (buf_size * cnt / msecs);
+  printf("DMA speed: %6ld kB/s (%d buffers of %d bytes in %ld msecs)\n", speed,
+         cnt, buf_size, msecs);
+  drive->silent++;
+  return 0;
+}
+
+int flush_cache(drive_info *drive, bool IMMED) {
+  //	drive->cmd[0] = SBC_FLUSH_CACHE;
+  drive->cmd[0] = MMC_SYNC_CACHE;
+  drive->cmd[1] = IMMED ? 0x02 : 0;
+
+  if ((drive->err = drive->cmd.transport(NONE, NULL, 0))) {
+    if (!drive->silent)
+      sperror("SBC_FLUSH_CACHE", drive->err);
+    return (drive->err);
+  }
+  //		{ sperror ("SBC_FLUSH_CACHE",drive->err); return (drive->err); }
+
+  return 0;
+}
+
+int set_cache(drive_info *drive, bool rd, bool wr) {
+  memset(drive->rd_buf, 0, 20);
+  drive->rd_buf[8] = MODE_PAGE_CACHING;
+  drive->rd_buf[9] = 0x0A;
+  drive->rd_buf[10] = (wr ? 0x04 : 0) | (rd ? 0 : 0x01);
+
+  return (mode_select(drive, 0x20));
+}
+
+int get_cache(drive_info *drive, bool *rd, bool *wr) {
+  bool re, we;
+  if (mode_sense(drive, MODE_PAGE_CACHING, 0, 20) ||
+      drive->rd_buf[8] != MODE_PAGE_CACHING)
+    return 1;
+  re = !(drive->rd_buf[10] & 0x01);
+  we = !!(drive->rd_buf[10] & 0x04);
+  printf("Cache:  RD %s, WR %s\n", re ? "EN" : "DIS", we ? "EN" : "DIS");
+  if (rd)
+    (*rd) = re;
+  if (wr)
+    (*wr) = we;
+  return 0;
+}
+
+int get_mode_pages_list(drive_info *drive) {
+  unsigned int len, i, ii;
```

Comment:
Codacy found an issue: [The scope of the variable 'ii' can be reduced.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)